lib/charms/identity_platform_login_ui_operator/v0/login_ui_endpoints.py

```diff
@@ -0,0 +1,179 @@
+        )
+
+    def _on_provider_endpoints_relation_created(self, event: RelationCreatedEvent) -> None:
+        self.on.ready.emit()
+
+    def send_endpoints_relation_data(self, endpoint: str) -> None:
+        """Updates relation with endpoint info."""
+        if not self._charm.unit.is_leader():
+            return
+
+        relations = self.model.relations[self._relation_name]
+        for relation in relations:
+            relation.data[self._charm.app].update(
+                {
+                    "consent": f"{endpoint}/consent",
+                    "error": f"{endpoint}/error",
+                    "index": f"{endpoint}/index",
+                    "login": f"{endpoint}/login",
+                    "oidc_error": f"{endpoint}/oidc_error",
+                    "registration": f"{endpoint}/registration",
+                    "browser": endpoint,
+                }
+            )
+
+
+class LoginUIEndpointsRelationError(Exception):
+    """Base class for the relation exceptions."""
+
+    pass
+
+
+class LoginUIEndpointsRelationMissingError(LoginUIEndpointsRelationError):
+    """Raised when the relation is missing."""
+
+    def __init__(self) -> None:
+        self.message = "Missing ui-endpoint-info relation with Identity Platform Login UI"
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRelationDataMissingError(LoginUIEndpointsRelationError):
+    """Raised when information is missing from the relation."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRequirer(Object):
+    """Requirer side of the ui-endpoint-info relation."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self._relation_name = relation_name
+
+    def get_login_ui_endpoints(self) -> Optional[Dict]:
+        """Get the Identity Platform Login UI endpoints."""
+        if not self.model.unit.is_leader():
+            return None
+        endpoints = self.model.relations[self._relation_name]
+        if len(endpoints) == 0:
+            raise LoginUIEndpointsRelationMissingError()
+
+        remote_app = [
+            app
+            for app in endpoints[0].data.keys()
+            if isinstance(app, Application) and not app._is_our_app
+        ][0]
+
+        data = endpoints[0].data[remote_app]
+        return_dict = {}
+
+        for k in RELATION_KEYS:
+            if k not in data:
+                raise LoginUIEndpointsRelationDataMissingError(
+                    "Missing endpoints in ui-endpoint-info relation data"
+                )
+            if data[k] == "":
+                raise LoginUIEndpointsRelationDataMissingError(
+                    "Missing endpoints in ui-endpoint-info relation data"
+                )
+            return_dict[k] = data[k]
```

Comment:
In relation interfaces we usually use the json schema that is uploaded to https://github.com/canonical/charm-relation-interfaces/ to validate the relation databag.

Let's try to keep that consistency.